Warehouse health integration

awisema opened the pull request:
Work in progress -- this is the submission file generator.

awisema committed:
Medicaid HMIS Interchange Schema

awisema committed:
Generate submission.

awisema committed:
Process error response.

awisema committed:
Error response should record affected external ids.

eanders commented on config/initializers/pluck_in_batches.rb:
```diff
@@ -27,7 +27,7 @@ def pluck_in_batches(columns, batch_size: 1_000)
       items = batch.pluck(*columns)
       page += 1
-      yield items
+      yield items if items.present?
```
What happens if items aren't present?  Seems odd we haven't hit this yet, but _maybe_ we've never plucked in batches with no results?

awisema replied:
I guess we haven't -- when I ran into it, it invoked the block on an empty array, which seemed wrong.

eanders commented on db/health/migrate/20230504194752_create_mhx_external_ids.rb:
```diff
@@ -0,0 +1,10 @@
+class CreateMhxExternalIds < ActiveRecord::Migration[6.1]
+  def change
+    create_table :mhx_external_ids do |t|
+      t.references :client
+      t.string :identifier
+      t.boolean :valid_id
+      t.timestamps
```
```suggestion
      t.references :client, null: false
      t.string :identifier, null: false,
      t.boolean :valid_id, null: false, default: true
      t.timestamps, null: false
```
alternatively, I wonder if `valid_id` should actually be a timestamp, `invalidated_at` and then we can put some logic around how long to wait to check again?

eanders replied:
Do we want a unique index on `client_id`, or maybe `client_id, identifier`?

eanders commented on drivers/medicaid_hmis_interchange/README.md:
```diff
@@ -0,0 +1,3 @@
+## MedicaidHmisInterchange README
+
+This README file should be used to explain the functionality of the driver.
```
```suggestion
Adds the ability to query MassHealth for MedicaidIDs and generates a file to assist in verifying homelessness of patients at MassHealth.
```

eanders commented on drivers/medicaid_hmis_interchange/app/models/medicaid_hmis_interchange/health/response.rb:
```diff
@@ -0,0 +1,44 @@
+        external_id = ExternalId.find_by(identifier: error[:medicaid_id])
+        if external_id.present?
+          external_id.update(valid_id: false)
+          external_ids << external_id
```
why are you adding this back to `external_ids`?  I think this is going to save on line 26, then create a duplicate record on 27.

awisema replied:
`external_ids` is the relation of `ExternalId`s that were flagged by MH. The `update` marks the id as invalid, and the shovel adds it to this `Response`. I didn't bother to batch this up, because I expect the error set to be very small.

eanders commented on drivers/medicaid_hmis_interchange/app/models/medicaid_hmis_interchange/health/submission.rb:
```diff
@@ -0,0 +1,122 @@
+      GrdaWarehouse::Hud::Client.homeless_on_date.pluck_in_batches(:id) do |batch|
+        lines = {}.tap do |results|
+          medicaid_ids = ExternalId.where(client_id: batch).
+            where.not(valid_id: false).
```
```suggestion
            where(valid_id: true).
```

eanders commented on drivers/medicaid_hmis_interchange/app/models/medicaid_hmis_interchange/health/submission.rb:
```diff
@@ -0,0 +1,122 @@
+    private def generate_filename(extension = 'txt', part = nil)
+      "rdc_homeless_#{part}#{@timestamp.strftime('%Y%m%d%H%M%S')}.#{extension}"
+    end
+
```
```suggestion

# Clients are included in the submission if they are enrolled in a homeless project (ES, SH, SO, TH) and not enrolled in PH with a move-in date in the past, on the day the process is run.
# When calculating length of time homeless, clients in NbN ES are given 30 days for each month they have at least one bed-night record, clients in SO are given 30 days for each month in which they have at least one Current Living Situation record.  In addition to the time in shelter or on the street time between DateToStreetESSH and EntryDate is added to the client's time homeless.
# For entry-exit enrollments, time between DateToStreetESSH and the Current date is counted toward time homeless.
```

eanders commented on drivers/medicaid_hmis_interchange/app/models/medicaid_hmis_interchange/health/submission.rb:
```diff
@@ -0,0 +1,122 @@
+            find_each do |client|
+            homeless_days = 0
+            client.service_history_enrollments.each do |enrollment|
+              if enrollment.nbn?
+                # 30 days for any month w/ service
+                months = enrollment.service_history_services.map(&:date).group_by(&:beginning_of_month).keys.count
+                homeless_days = [
+                  homeless_days,
+                  months * 30,
+                ].max
+              elsif enrollment.so?
+                # 30 days for any month w/ CLS
+                months = enrollment.enrollment.current_living_situations.map(&:InformationDate).group_by(&:beginning_of_month).keys.count
+                homeless_days = [
+                  homeless_days,
+                  months * 30,
+                ].max
```
These two also need time between DateToStreetESSH and EntryDate added.

awisema committed:
Exchange workflow.

awisema committed:
Spec 1.3 changes.

awisema committed:
Added gems for MH sftp.

awisema committed:
Update drivers/medicaid_hmis_interchange/README.md

Co-authored-by: Elliot <elliot@greenriver.org>

awisema committed:
Changes from review.
- Record id invalidation date instead of just status.
- Include pre-service days in time homeless for NbN/SO.

eanders reviewed (approved):
This looks great to me.  Thanks for the re-work.

awisema committed:
Scheduled rake task to deliver file.

awisema committed:
270/271s to retrieve medicaid ids.

awisema committed:
Schedule medicaid id lookups.

awisema committed:
Disable SFTP tests

awisema committed:
Merge branch 'release-74' into aw/warehouse-health-integration-184574932

eanders committed:
Merge branch 'release-74' into aw/warehouse-health-integration-184574932

awisema commented on drivers/medicaid_hmis_interchange/app/jobs/medicaid_hmis_interchange/medicaid_id_lookup_job.rb:
```diff
@@ -0,0 +1,45 @@
+
+    def perform(client_ids, test: false)
+      soap = ::Health::Soap::MassHealth.new(test: test)
+      return unless soap.configured?
```
This needs another guard -- otherwise, it will end up getting run in BHCHP.

eanders replied:
@awisema are you adding the guard now?  I just pushed a possible fix for cron syntax generation (guess we didn't have any weekly yet.)

eanders committed:
Attempt to send what AWS wants for day-based crons

eanders committed:
Attempting to fix cron syntax for aws

eanders commented on lib/tasks/health.rake:
```diff
@@ -230,6 +230,21 @@ namespace :health do
     end
   end
 
+  desc "Upload Medicaid homelessness information"
+  task medicaid_hmis_transfer: [:environment, 'log:info_to_stdout'] do
+    MedicaidHmisInterchange::FileExchangeJob.perform_later
+  end
+
+  desc "Query for medicaid ids"
+  task medicaid_id_query: [:environment, 'log:info_to_stdout'] do
+    homeless_clients = GrdaWarehouse::Hud::Client.homeless_on_date
+    homeless_clients.where.not(
+      id: MedicaidHmisInterchange::Health::ExternalId.pluck(:identifier), # TODO: Re-process invalidated identifiers?
+    ).pluck_in_batches(:id, batch_size: 100) do |batch|
+      MedicaidHmisInterchange::MedicaidIdLookupJob.perform_later(batch)
+    end
+  end
```
I missed this, but this can totally go in the drivers `lib/tasks` directory.

awisema committed:
Bugfixes.

awisema committed:
Don't run medicaid id lookup if MHX is disabled.

awisema committed:
Bugfix.

awisema committed:
Move rake tasks into driver.

eanders committed:
Fixes for lookups

eanders committed:
Use key file instead of key value

eanders committed:
Require SHE to have enrollments

eanders committed:
More MassHealth and SFPT tweaks

eanders committed:
Preload Current Living Situations

eanders committed:
Fix incorrect CLS preload

eanders committed:
Limit client scope to those who have medicaid ids

eanders committed:
Fix for zip spec with join to enrollment; cleanup old should code

eanders committed:
SFTP tests, adjust to multiple directory support

eanders committed:
Attempting to get SFTP working in CI

eanders committed:
Attempting to get SFTP working in CI

eanders committed:
Attempting to get SFTP working in CI

eanders committed:
Attempting to get SFTP working in CI

eanders committed:
Fixes for metadata file; test file data; duplicate removal

eanders committed:
Use same timestamp format for metadata and filename

eanders committed:
Merge branch 'release-74' into aw/warehouse-health-integration-184574932

eanders committed:
Adjustments for processing error report and additional test file

eanders committed:
Test data option

eanders committed:
Fix backwards check

eanders committed:
Test updates; new response headers

eanders committed:
Merge branch 'release-74' into aw/warehouse-health-integration-184574932
